### PR TITLE
AMKO docs for 1.4.1

### PR DESCRIPTION
**Pull request by hemantavi**

### CODE REVIEW

File: docs/AMKO/README.md [hunk 1]
```diff
@@ -74,28 +74,28 @@ The extra `update` permission is to update the `GSLBConfig` and `GlobalDeploymen
...
-   $ helm install  ako/amko  --generate-name --version 1.3.1 -f /path/to/values.yaml  --set configs.gsllbLeaderController=<leader_controller_ip> --namespace=avi-system
+   $ helm install  ako/amko  --generate-name --version 1.4.1 -f /path/to/values.yaml  --set configs.gsllbLeaderController=<leader_controller_ip> --namespace=avi-system
    ```
 6. Check the installation:
    ```
    $ helm list -n avi-system
 
    NAME           	NAMESPACE 	REVISION	UPDATED                                	STATUS  	CHART                 	APP VERSION
-   amko-1598451370	avi-system	1       	2020-08-26 14:16:21.889538175 +0000 UTC	deployed	amko-1.3.1	            1.3.1
+   amko-1598451370	avi-system	1       	2020-08-26 14:16:21.889538175 +0000 UTC	deployed	amko-1.3.1	            1.4.1
```
Comment (RiyaGaniga-avi) [inline docs/AMKO/README.md:98]: Shouldn't it be 1.4.1 in the chart column

Comment (hemantavi) [reply docs/AMKO/README.md]: Yes. Rather `1.4.1-beta`. Fixed, thanks!

File: docs/AMKO/crds/gslbhostrule.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+A CRD called `GSLBHostRule` allows users to override certain properties of the GslbServices on the Avi Controller created by AMKO. A typical `GSLBHostRule` looks like this:
+```yaml
+apiVersion: amko.vmware.com/v1alpha1
+kind: GSLBHostRule
+metadata:
+  name: gslb-host-rule-1
+  namespace: avi-system
+spec:
+  fqdn: foo.avi.internal
+  sitePersistence:
```
Comment (RiyaGaniga-avi) [inline docs/AMKO/crds/gslbhostrule.md:11]: sitePersistence is not an array but an object. Right?

Comment (hemantavi) [reply docs/AMKO/crds/gslbhostrule.md:11]: Yes, fixed. Thanks!

File: docs/AMKO/README.md [hunk 3]
```diff
@@ -149,9 +154,30 @@ NAME         AGE
...
 #### Editing runtime parameters of AMKO
 The `GDP` object can be edited at runtime to change the application selection parameters, traffic split and the applicable clusters. AMKO will recognize these changes and will update the GSLBServices accordingly.
 
 Changing only `logLevel` is permissible at runtime via an edit of the `GSLBConfig`. For all other changes to the `GSLBConfig`, the AMKO pod has to be restarted.
+
+#### Choosing a mode of GslbService FQDN
+There can be different requirements for a user to either use local FQDNs as the GslbService FQDNs, or use a different FQDN as the Global FQDN. Please see [this](docs/local_and_global_fqdn.md) to choose a mode fit for the use-case and enable accordingly.
+
+#### Gslb Service Properties
+Certain Gslb Service properties can be set and modified at runtime. If these are set through the GDP object, they are applied to all the Gslb Services. If a user wants to override specific properties, they can use a `GSLBHostRule` object for a GslbService.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:173]: This is giving a feeling that both modes co-exist together. But aren't me making them exclusive of one another?

File: docs/AMKO/crds/gdp.md [hunk 4]
```diff
@@ -76,6 +87,17 @@ matchRules:
 
 4. `trafficSplit` is required if we want to route a percentage of traffic to objects in a given cluster. Weights for these clusters range from 1 to 20.
 
+5. `ttl`: Time To Live, ranging from 1-86400 seconds that determines the frequency with which clients need to obtain fresh steering information for client requests. If none is specified in the GDP object, the value defaults to the one specified in the DNS application profile.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:90]: Can we reword this:
Time To Live, ranging from 1-86400 seconds that determines the frequency with which clients need to obtain fresh steering information for client requests.

Use this flag to set the time to live value. The value can range from 1-86400 seconds. This determines the frequency with which clients need to obtain fresh steering information for client requests

File: docs/AMKO/crds/gslbhostrule.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+2. `fqdn`: FQDN of the GslbService.
+
+3. `sitePersistence`: Enable Site Persistence for client requests. Set the `enabled` flag as `true` and add a `profileRef` for a pre-created Application Persistence Profile created on the Avi Controller. Please follow the steps [here](https://avinetworks.com/docs/20.1/gslb-site-cookie-persistence/#outline-of-steps-to-be-taken) to create a federated Application Persistence Profile on the Avi Controller.
+
+**Note** that site persistence is **disabled** on GslbServices created for **insecure** ingresses/routes, irrespective of this field.
+If this field is not provided in `GSLBHostRule`, the site persistence property will be inherited from the GDP object.
+
+4. `thirdPartyMembers`: To add some third party members to a GS from a non-avi site (third party site) for the purpose of maintenance, specify a list of those members. For each member, provide the site name in `site` and IP address in `vip`.
+
+**Note** that the site must be added to the GSLB leader as a 3rd party site before adding the member here.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:38]: Can we refer an AVI KB on how third party sites are added?

File: docs/AMKO/crds/gslbhostrule.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+## GSLBHostRule CRD for AMKO
+A CRD called `GSLBHostRule` allows users to override certain properties of the GslbServices on the Avi Controller created by AMKO. A typical `GSLBHostRule` looks like this:
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:2]: The `GSLBHostRule` CRD allows users to override certain properties of a specific GSLB object on the Avi Controller created by AMKO.

File: docs/AMKO/crds/gslbhostrule.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    weight: 15
+  - cluster: oshift
+    weight: 5
+  ttl: 30
+```
+1. `namespace`: namespace of this object must be `avi-system`.
+
+2. `fqdn`: FQDN of the GslbService.
+
+3. `sitePersistence`: Enable Site Persistence for client requests. Set the `enabled` flag as `true` and add a `profileRef` for a pre-created Application Persistence Profile created on the Avi Controller. Please follow the steps [here](https://avinetworks.com/docs/20.1/gslb-site-cookie-persistence/#outline-of-steps-to-be-taken) to create a federated Application Persistence Profile on the Avi Controller.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:31]: I think what is confusing to me reading this section is the following:
- what happens when user specifies `enabled` to `false` and specifies a `profileRef`
- what happens if the user specifies just the `enabled` flag to `true` and does not specify the `profileRef`

Basically I am not clear on why the enabled flag is used here. Can't we just have the `profileRef` specified, that would automatically indicate that the user wants to override the default settings?

File: docs/AMKO/crds/gslbhostrule.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+**Note** that the site must be added to the GSLB leader as a 3rd party site before adding the member here.
+
+5. `healthMonitorRefs`: If a GslbService requires some custom health monitoring, the user can create a federated custom health monitor in the Avi Controller and provide the ref(s) here. To add a custom health monitor, follow the steps [here](https://avinetworks.com/docs/16.3/avi-gslb-service-and-health-monitors/#configuring-health-monitoring). If no custom health monitor refs have been added, the custom health monitors from the GDP object will be inherited.
+
+6. `trafficSplit`: Specify traffic steering to member clusters/sites. The traffic is then split proportionately between two different clusters. Weight for each cluster must be provided between 1 to 20. If not added, GDP object's traffic split applies on this GslbService.
+
+7. `ttl`: Time To Live, specify how much time does a resolver keep the A Record for an FQDN. If not added, `ttl` specified on the GDP object is inherited for this GslbService.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:44]: Can we reword to:
Override the default `ttl` value specified on the GDP object using this field.

Trying to keep it simple.

File: docs/AMKO/crds/gslbhostrule.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+## Caveats:
+1. Site Persistence isn't enabled for the GslbServices which have insecure ingresses or routes as the members.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:47]: Site persistence cannot be enabled for...

File: docs/AMKO/crds/gslbhostrule.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+2. If any of the properties in this object isn't specified, the respective property is inherited from the GDP object.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbhostrule.md:48]: This is understood from the beginning of the section itself.

File: docs/AMKO/docs/local_and_global_fqdn.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# Deriving GslbService FQDNs
+
+## FQDN Modes
+AMKO decides the FQDNs for the GslbServices based on two modes:
+1. Default Mode: Here, the hostname fields in the ingress, routes and service type load balancer become the deciding factor. If an ingress gets added with a hostname (and if its also selected via the GDP object), AMKO takes the hostname(s) present in the status field of the ingress and creates GslbService(s) with the respective hostnames. In this mode, no other objects are used to derive the mapping.
```
Comment (sudswasavi) [inline docs/AMKO/docs/local_and_global_fqdn.md:5]: Can we re-tune this section:

In this mode, the hostname(s) field in the status of the Ingress/Route/Service of type Loadbalancer object is used to determine the hostname of the GSLB Service. Each hostname uniquely maps to a GS FQDN automatically. For common hostname across clusters, a single GSLB Service is created with pool members from each cluster that share the hostname.

File: docs/AMKO/docs/local_and_global_fqdn.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+2. Custom Global Fqdn Mode: In this mode, AMKO checks the AKO HostRules to figure out the GslbService Fqdn. To expose an application via GSLB, the user must provide a mapping between the local fqdn and the global fqdn via AKO's HostRule object. See [this](https://github.com/avinetworks/avi-helm-charts/blob/master/docs/AKO/crds/hostrule.md#configure-gslb-fqdn) to configure a mapping between the local FQDN and a global FQDN. If the user wishes to map two application instances (which have different FQDNs) in two different clusters, they have to create HostRules in both clusters specifying the mapping.
```
Comment (sudswasavi) [inline docs/AMKO/docs/local_and_global_fqdn.md:7]: to configure a mapping between the local FQDN and a global FQDN. --> this line can be removed.

File: docs/AMKO/docs/local_and_global_fqdn.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+![Alt text](../images/local_vs_global_fqdn.png?raw=true "local and global fqdn modes")
+
+## How to specify the custom global fqdn mode
+The user has to set the `useCustomGlobalFqdn` to `true` in the `GSLBConfig` object. This is a static operation and if changed while AMKO is already deployed, the changes won't take any effect.
```
Comment (sudswasavi) [inline docs/AMKO/docs/local_and_global_fqdn.md:12]: What happens if this mode is set to `true` and AMKO is rebooted while it had objects pre-created in the Avi Controller using local FQDNs?

File: docs/AMKO/docs/local_and_global_fqdn.md [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+## When to use the default mode
+![Alt text](../images/local_fqdn.png?raw=true "local fqdn usage")
+
+As mentioned, in the default mode, AMKO just looks at the ingress, routes and LB services and their hostnames to derive the FQDNs for the GslbServices. Users will find it useful when they have two instances of an application deployed on two different clusters which have the same FQDNs. And, they would want to expose all these applications to be exposed as GslbServices in one shot.
```
Comment (sudswasavi) [inline docs/AMKO/docs/local_and_global_fqdn.md:17]: As mentioned, in the default mode, AMKO just looks at the ingress, routes and LB services and their hostnames to derive the FQDNs for the GslbServices. --> can be removed.

Use this mode if two instances of an application deployed on two different clusters have the same FQDN across clusters.

File: docs/AMKO/docs/local_and_global_fqdn.md [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+## When to use Custom Global Fqdn mode
+![Alt text](../images/global_fqdn.png?raw=true "global fqdn usage")
+
+For users who have specific site local FQDNs for their applications and these FQDNs are not common for applications deployed on two different clusters (sites), they can deploy AMKO in Custom Global Fqdn Mode. Users must create HostRule mapping their local FQDNs to a global FQDN in their respective clusters. AMKO then patches both the applications to the same GslbService with global FQDN.
```
Comment (sudswasavi) [inline docs/AMKO/docs/local_and_global_fqdn.md:22]: This entire section can be re-worded to:

If users have site local FQDNs and they would want to use DNS loadbalancing for these application instances a common GSLB FQDN can be used. Here the common GSLB FQDN maps the vips of the site local FQDNs as pool members.

**Review by sudswasavi — APPROVED** [review]

**Pull request merged**